Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h

```diff
@@ -54,38 +84,53 @@ class WebXROpaqueFramebuffer {
     static std::unique_ptr<WebXROpaqueFramebuffer> create(PlatformXR::LayerHandle, WebGLRenderingContextBase&, Attributes&&, IntSize);
     ~WebXROpaqueFramebuffer();
 
+    bool supportsDynamicViewportScaling() const;
+
     PlatformXR::LayerHandle handle() const { return m_handle; }
-    const WebGLFramebuffer& framebuffer() const { return m_framebuffer.get(); }
-    GCGLint width() const { return m_framebufferSize.width(); }
-    GCGLint height() const { return m_framebufferSize.height(); }
+    const WebGLFramebuffer& framebuffer() const { return m_drawFramebuffer.get(); }
+    IntSize displayFramebufferSize() const;
+    IntSize drawFramebufferSize() const;
+    IntRect drawViewport(PlatformXR::Eye) const;
 
     void startFrame(const PlatformXR::FrameData::LayerData&);
     void endFrame();
 
 private:
     WebXROpaqueFramebuffer(PlatformXR::LayerHandle, Ref<WebGLFramebuffer>&&, WebGLRenderingContextBase&, Attributes&&, IntSize);
 
-    bool setupFramebuffer();
-    PlatformGLObject allocateRenderbufferStorage(GraphicsContextGL&, GCGLsizei, GCGLenum, IntSize);
-    PlatformGLObject allocateColorStorage(GraphicsContextGL&, GCGLsizei, IntSize);
-    PlatformGLObject allocateDepthStencilStorage(GraphicsContextGL&, GCGLsizei, IntSize);
-    void bindColorBuffer(GraphicsContextGL&, PlatformGLObject);
-    void bindDepthStencilBuffer(GraphicsContextGL&, PlatformGLObject);
+#if PLATFORM(COCOA)
+    bool setupFramebuffer(GraphicsContextGL&, const PlatformXR::FrameData::LayerSetupData&);
+#endif
+    void allocateRenderbufferStorage(GraphicsContextGL&, GCGLOwnedRenderbuffer&, GCGLsizei, GCGLenum, IntSize);
+    void allocateAttachments(GraphicsContextGL&, WebXRAttachments&, GCGLsizei, IntSize);
+    void bindAttachments(GraphicsContextGL&, WebXRAttachments&);
+    void resolveMSAAFramebuffer(GraphicsContextGL&);
+    void blitShared(GraphicsContextGL&);
+    void blitSharedToLayered(GraphicsContextGL&);
+    IntRect calculateViewportShared(PlatformXR::Eye, bool, const IntRect&, const IntRect&);
 
     PlatformXR::LayerHandle m_handle;
-    Ref<WebGLFramebuffer> m_framebuffer;
+    Ref<WebGLFramebuffer> m_drawFramebuffer;
     WebGLRenderingContextBase& m_context;
     Attributes m_attributes;
+    PlatformXR::Layout m_displayLayout = PlatformXR::Layout::Layered;
```

Comment:
I don't think should default to `Layered`. I'd prefer to have the init code detect and flip it.

Reply:
Oh we do detect already, so I can probably just revert this to Shared